subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java

```diff
@@ -81,23 +87,56 @@ private void registerSoftwareTypes(TypeMetadata typeMetadata) {
+        softwareTypePluginImplMetadata.visitValidationFailures(null, typeValidationContext);
 
-        for (PropertyMetadata propertyMetadata : softwareTypePluginImplMetadata.getPropertiesMetadata()) {
-            Optional<SoftwareType> registersSoftwareType = propertyMetadata.getAnnotation(SoftwareType.class);
-            if (registersSoftwareType.isPresent()) {
-                return;
-            }
+        long exposedSoftwareTypes = softwareTypePluginImplMetadata.getPropertiesMetadata().stream()
+            .map(propertyMetadata -> propertyMetadata.getAnnotation(SoftwareType.class))
+            .filter(Optional::isPresent)
+            .count();
+
+        if (exposedSoftwareTypes == 0) {
+            typeValidationContext.visitTypeProblem(problem ->
+                problem.withAnnotationType(softwareTypePluginImplClass)
+                    .id("missing-software-type", "Missing software type annotation", GradleCoreProblemGroup.validation().type())
+                    .contextualLabel("is registered as a software type plugin but does not expose a software type")
+                    .severity(Severity.ERROR)
+                    .details("This class was registered as a software type plugin, but it does not expose a software type. Software type plugins must expose exactly one software type via a property with the @SoftwareType annotation.")
+                    .solution("Add @SoftwareType annotations to properties of " + softwareTypePluginImplClass.getSimpleName())
+                    .solution("Remove " + softwareTypePluginImplClass.getSimpleName() + " from the @RegistersSoftwareTypes annotation on " + registeringPlugin.getSimpleName())
+            );
+        } else if (exposedSoftwareTypes > 1) {
+            typeValidationContext.visitTypeProblem(problem ->
+                problem.withAnnotationType(softwareTypePluginImplClass)
+                    .id("missing-software-type", "Missing software type annotation", GradleCoreProblemGroup.validation().type())
+                    .contextualLabel("is registered as a software type plugin, but it exposes multiple software types")
+                    .severity(Severity.ERROR)
+                    .details("This class was registered as a software type plugin, but it exposes multiple software types. Software type plugins must expose exactly one software type via a property with the @SoftwareType annotation.")
+                    .solution("Add the @SoftwareType annotation to only one property of " + softwareTypePluginImplClass.getSimpleName())
+            );
         }
 
-        throw new InvalidUserDataException("A plugin with type '" + softwareTypePluginImplClass.getName() + "' was registered as a software type plugin, but it does not expose any software types. Software type plugins must expose software types via properties with the @SoftwareType annotation.");
+        if (!typeValidationContext.getProblems().isEmpty()) {
+            throw new DefaultMultiCauseException(
```

Comment:
🤔 Can't you only ever get a single validation problem out of this code?  Why start with `MultiCause`?

Reply:
A few lines up, we call `softwareTypePluginImplMetadata.visitValidationFailures()` which captures any validation failures it encountered when inspecting the type.  Right now this only includes the validations in `SoftwareTypeAnnotationHandler` which checks that the public type is a super class of (or the same class as) the method return type.  So you could have say, mismatched types _and_ multiple exposed types.